### PR TITLE
Updated olefile types

### PR DESCRIPTION
**Pull request by radarhere**
1. `_parse_property()` [returns `v`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L2245), which might be [the first item from the tuple returned by `_parse_property_basic()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L2219), which is `int | str | bytes | bool | None`
https://github.com/python/typeshed/blob/6220c20d9360b12e2287511587825217eec3e5b5/stubs/olefile/olefile/olefile.pyi#L229-L231
So those are additional possible return types.
2. That then [affects the return type of `getproperties()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L2205-L2214)
3. For `OleDirectoryEntry`, I've added `clsid`, which is set in [`__init__()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L839) to [`str`](https://github.com/python/typeshed/blob/6220c20d9360b12e2287511587825217eec3e5b5/stubs/olefile/olefile/olefile.pyi#L129)
4. For `OleFileIO`, I've added `root`, which is [set to `None` in `__init__()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1117), and then [later set](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1801-L1803) to [an instance of `OleDirectoryEntry`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1833)

### CODE REVIEW

File: stubs/olefile/olefile/olefile.pyi [hunk 1]
```diff
@@ -222,10 +225,10 @@ class OleFileIO:
     def get_rootentry_name(self) -> bytes: ...
     def getproperties(
         self, filename: str | Sequence[str], convert_time: bool = False, no_conversion: list[int] | None = None
-    ) -> dict[int, list[int | str | bytes | bool | None]]: ...
+    ) -> dict[int, list[int | str | bytes | bool | None] | int | str | bytes | bool | None]: ...
     def _parse_property(
         self, s: bytes, offset: int, property_id: int, property_type: int, convert_time: bool, no_conversion: list[int]
-    ) -> list[int | str | bytes | bool | None] | None: ...
+    ) -> list[int | str | bytes | bool | None] | int | str | bytes | bool | None: ...
```
Comment (sobolevn) [inline stubs/olefile/olefile/olefile.pyi:231]: looks like a good candidate for an alias.

Comment (radarhere) [reply stubs/olefile/olefile/olefile.pyi]: I've added a commit.

**Comment by radarhere** [comment]
I've also realised that based on the [path encoding](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1071-L1074) of the OleFileIO instance, `_decode_utf16_str()` might return a string or bytes, meaning that the OleDirectoryEntry [`name`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L818C30-L818C48) might be string or bytes, affecting [`_list()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1879-L1890) and [`listdir()`](https://github.com/decalage2/olefile/blob/ae8e110e90965f2648052b29f1b3da2c39465001/olefile/olefile.py#L1907-L1918)

**Comment by srittau** [comment]
I have no experience with olefile, but looking at the source code, it seems clear that it handles both `str` and `bytes` paths, depending on whether `OleFileIO.path_encoding` is `None` or not. That said, it seems dangerous to me to just change `str` to `str | bytes`. Making the classes generic of the string type seems like a much safer approach to me. But again: I have no experience with olefile, so this might not make sense or be impossible.

**Comment by radarhere** [comment]
> Making the classes generic of the string type seems like a much safer approach to me.

Sorry, do you mean `AnyStr`?

**Comment by srittau** [comment]
Generic over the string type, using `AnyStr`, yes.

**Comment by radarhere** [comment]
Ok, I've made that change.

File: stubs/olefile/olefile/olefile.pyi [hunk 2]
```diff
@@ -122,7 +122,7 @@ DEFECT_INCORRECT: int
 DEFECT_FATAL: int
 MINIMAL_OLEFILE_SIZE: int
 
-def isOleFile(filename: IO[bytes] | bytes | str | None = None, data: bytes | None = None) -> bool: ...
+def isOleFile(filename: IO[bytes] | AnyStr | None = None, data: bytes | None = None) -> bool: ...
```
Comment (srittau) [inline stubs/olefile/olefile/olefile.pyi:125]: In this case, `AnyStr` doesn't do anything. It's a type variable, and is appears only once in this definition, so using `bytes | str` was correct here.

File: stubs/olefile/olefile/olefile.pyi [hunk 3]
```diff
@@ -172,10 +173,14 @@ class OleDirectoryEntry:
     def getmtime(self) -> datetime.datetime | None: ...
     def getctime(self) -> datetime.datetime | None: ...
 
+_Property: TypeAlias = int | str | bytes | bool | None
+
 class OleFileIO:
```
Comment (srittau) [inline stubs/olefile/olefile/olefile.pyi:178]:
```suggestion
class OleFileIO[Generic[AnyStr]:
```

One problem I see is that the type var can't be determined statically from `__init__`, unless `bytes` or `str` is used. So we will need an overload for the `IO[bytes]` and `None` cases.

**Review by srittau — CHANGES_REQUESTED** [review]
Partial review below.

File: stubs/olefile/olefile/olefile.pyi (hunk 2 above)
Comment (radarhere) [reply stubs/olefile/olefile/olefile.pyi]: Ok, I've pushed a commit for this.

File: stubs/olefile/olefile/olefile.pyi [hunk 4]
```diff
@@ -153,16 +153,16 @@ class OleStream(io.BytesIO):
         sectorsize: int,
         fat: list[int],
         filesize: int,
-        olefileio: OleFileIO,
+        olefileio: OleFileIO[AnyStr],
     ) -> None: ...
 
 class OleDirectoryEntry:
```
Comment (srittau) [inline stubs/olefile/olefile/olefile.pyi:159]: I guess this also needs to be generic.

**Review by srittau — APPROVED** [review]
Thanks!

**Pull request merged**